### PR TITLE
This PR addresses issue #831, adding public Router#resolve method 

### PR DESCRIPTION
**Pull request by hogart**
Along with necessary typings and docs.

From technical perspective this change adds protected `_resolve` and public `resolve` methods, latter being wrapper around first. `_resolve` is used in `<router-link/>`, since it requires a lot of additional info. `resolve` just returns relevant bits from that.

Function `createHref` was moved to `src/index.js`, but maybe belongs in `sc/util/path.js`.

### CODE REVIEW

**Comment by posva** [comment]
Thanks for the pr!
I'd rather have a single resolve method. Let's wait for @fnlctrl and @yyx990803 input

File: src/components/link.js [hunk 1]
```diff
@@ -26,14 +24,10 @@ export default {
   render (h: Function) {
     const router = this.$router
     const current = this.$route
-    const to = normalizeLocation(this.to, current, this.append)
-    const resolved = router.match(to)
-    const fullPath = resolved.redirectedFrom || resolved.fullPath
-    const base = router.history.base
-    const href = createHref(base, fullPath, router.mode)
+    const { normalizedTo, resolved, href } = this.$router._resolve(this.to, current, this.append)
```
Comment (posva) [inline src/components/link.js]: you can directly call `router._resolve`

**Comment by hogart** [comment]
> I'd rather have a single resolve method.

Me too. We can rename that internal method, I think.

**Comment by LinusBorg** [comment]
I like this PR. @hogart, Can you solve the conflicts?

**Comment by hogart** [comment]
Thanks, @LinusBorg. I merged my branch with upstream.

**Comment by LinusBorg** [comment]
@hogart sorry to bother again. This also needs a a unnit test for the new #resolve() function. Could you add one?

**Comment by fnlctrl** [comment]
Sorry I forgot this... I agree with @posva that we should use a single resolve method, both internally and as public api. Let's wait for Evan's input on this.

**Comment by hogart** [comment]
@LinusBorg I'd like to, but I didn't found place for testing Router class.

@fnlctrl, @posva how do you imagine it? We need several pieces of info in `link` component, most of which are useless outside of that context. On the other hand, resolving algorithm stays the same, so it's either this or duplicating code.

**Comment by posva** [comment]
@hogart By one method I meant that I'd rather have `_getResolutionData` inside `resolve` and use resolve everywhere

**Comment by hogart** [comment]
@posva I understand what you want, but, sorry, I don't see how we can use `resolve` in `link` component, because it requires additional data. 
Can you please elaborate how in your vision it should work?

**Comment by posva** [comment]
@hogart Oh, I wanted to say that resolve would return the whole object not just an `href` 😆

**Comment by hogart** [comment]
@posva that's definitely possible:) But in my opinion this pollutes API surface, are you guys sure about that?

**Comment by posva** [comment]
no, I personally think it's ok. Why does it pollutes the api surface, is it because you may never use the rest of the returned object?

**Comment by hogart** [comment]
@posva yes, because of that.
Ok, I implement appropriate changes. Anything to land this PR already:)

**Comment by hogart** [comment]
I'm closing this PR because messed-up changelog, see #918 instead.

**Pull request closed without merging**